leader-election-impl/src/test/java/com/palantir/paxos/PaxosStateLogMigratorTest.java

```diff
@@ -66,75 +81,104 @@ public void emptyLogMigrationSuccessfullyMarksAsMigrated() {
     }
 
     @Test
-    public void logMigrationSuccessfullyMigratesEntries() {
+    public void logMigrationWithNoLowerBoundMigratesOnlyGreatest() {
         long lowerBound = 10;
         long upperBound = 25;
-        List<PaxosValue> valuesWritten = insertValuesWithinBounds(lowerBound, upperBound, source);
+        insertValuesWithinBounds(lowerBound, upperBound, source);
 
-        migrateFrom(source);
+        long cutoff = migrateFrom(source);
+
+        assertThat(cutoff).isEqualTo(upperBound);
         assertThat(migrationState.hasMigratedFromInitialState()).isTrue();
-        assertThat(target.getLeastLogEntry()).isEqualTo(lowerBound);
+        assertThat(migrationState.isInMigratedState()).isTrue();
+        assertThat(target.getLeastLogEntry()).isEqualTo(upperBound);
         assertThat(target.getGreatestLogEntry()).isEqualTo(upperBound);
 
-        valuesWritten.forEach(value -> assertThat(getPaxosValue(target, value.seq)).isEqualTo(value));
+        LongStream.rangeClosed(lowerBound, upperBound - 1)
+                .mapToObj(sequence -> readRoundUnchecked(target, sequence))
+                .map(Assertions::assertThat)
+                .forEach(AbstractAssert::isNull);
+        assertThat(getPaxosValue(target, upperBound)).isEqualTo(valueForRound(upperBound));
+    }
+
+    @Test
+    public void logMigrationWithLowerBoundMigratesFromBound() {
+        long lowerBound = 10;
+        long upperBound = 25;
+        insertValuesWithinBounds(lowerBound, upperBound, source);
+
+        long expectedCutoff = 17;
+        long cutoff = migrateFrom(source, OptionalLong.of(expectedCutoff));
+
+        assertThat(cutoff).isEqualTo(expectedCutoff);
+        assertThat(migrationState.hasMigratedFromInitialState()).isTrue();
+        assertThat(migrationState.isInMigratedState()).isTrue();
+        assertThat(target.getLeastLogEntry()).isEqualTo(expectedCutoff);
+        assertThat(target.getGreatestLogEntry()).isEqualTo(upperBound);
+
+        LongStream.rangeClosed(lowerBound, expectedCutoff - 1)
+                .mapToObj(sequence -> readRoundUnchecked(target, sequence))
+                .map(Assertions::assertThat)
+                .forEach(AbstractAssert::isNull);
+        KeyedStream.of(LongStream.rangeClosed(expectedCutoff, upperBound).boxed())
+                .map(sequence -> getPaxosValue(target, sequence))
+                .mapKeys(PaxosStateLogTestUtils::valueForRound)
+                .entries()
+                .forEach(entry -> assertThat(entry.getKey()).isEqualTo(entry.getValue()));
     }
```

Comment:
Let's document the behaviour where `expectedCutoff > upperBound` (which currently is not migrating anything, which is weird but we may not want to ban that).

Reply:
Yeah, this actually doesn't play nice with splitting log so changed